### ISSUE TITLE
Losing touch input after a suspend...

### ISSUE THREAD

**Issue by NiLuJe** [koreader/koreader#1862]
* KOReader version: 2015.11-185, 2015.11-183, 2015.11-175
* Device: Kobo H2O

#### Issue

I appear to be losing touch input after a wakeup from suspend...
I don't recall this ever happening before, but it's been a while...
That's on FW 3.19.5761

#### Steps to reproduce

Suspend.
Wakeup.
Enjoy your hard-reboot unless you're still in USBNet mode ;).

----

I though this was an USBNet side-effect, because I had it randomly happen on nickel too, but that might have been *after* a broken wakeup from KO...
And, anyway, it happens without USBNetwork, too :/.

Keep in mind that I'm not using KSM, and launching KOReader manually in a shell (and backgrounding it if I'm over WiFi).

Do we need to ungrab & close/open & grab event1 around suspend/wakeup?
We currently can't, we only have a closeAll function, and we don't keep a map of open fd to input device...

**Comment by NiLuJe** [comment]
Setting gSleep_Mode_Suspend to 1 (echo 1 >| /sys/power/state-extended) and restarting nickel fixes it...

Which tells me that something's fishy going in with how the Kernel handles suspend/wakeup..
-_-"

**Comment by NiLuJe** [comment]
In particular, I'm not seeing 

````
[drivers/input/touchscreen/zforce_i2c.c-671] zforce_i2c_open()
[zForce_ir_touch_recv_data-225] command Activate (0) ...
[zForce_ir_touch_recv_data-234] command Resolution (0) ...
[zForce_ir_touch_recv_data-268] command Frequency (0) ...
[zForce_ir_touch_recv_data-243] command Configuration ...
````

In dmesg upon wakeup when this happens (be it in nickel or KOReader).

**Comment by NiLuJe** [comment]
Until we get our mitts in its guts, nickel doesn't even try to suspend when pressing the power button when plugged in w/ USBNet enabled...

Why that checks gets chucked through the window once we restart it is a mystery...

That, and all the other crap relating to putting stuff in low power and/or resuming that appears in dmesg during suspend/wakeup cycles when we leave nickel alone and that we never ever see when we do things (badly) ourselves.

**Comment by NiLuJe** [comment]
TL;DR: I don't particularly care about this, but, clearly, we're missing something.

**Comment by NiLuJe** [comment]
In fact, I don't even know what the hell is actually supposed to resume the device.

That's certainly not the ```echo 0 > /sys/power/state-extended``` since that just sets a global int to 0 in the kernel (unless the kernel sources are outdated/wrong, which is a decent possibility, since I'm seeing stuff in dmesg that's nowhere in those sources...), and commenting it yields the exact same behavior.

**Comment by houqp** [comment]
That's exactly what I feel while reading the kernel source a while ago. That store function makes no sense to me >_<.

@lgeek Do you know how their suspend code works?

**Comment by NiLuJe** [comment]
Okay, dirty workaround:

```echo "a" >| /sys/devices/virtual/input/input1/neocmd``` wakes up the IR grid.

**Comment by NiLuJe** [comment]
So, open question to Kobo users w/ devices w/ an IR grid: am I the only one suffering from this issue?

**Comment by Frenzie** [comment]
I seem to be on -161 and haven't noticed touch not working after resume on
H2O.

**Comment by NiLuJe** [comment]
@Frenzie: On which FW version? And how do you launch KOReader?

**Comment by KenMaltby** [comment]
Just tried -194 on my AuraHD, with FW 3.0.5, launched using KSM7 - touch still working after return from suspend/sleep.

**Comment by AlanSP1** [comment]
I don't have problems with suspending device and touch screen.

H2O, FW 3.12, KSM8, Koreader -175, CoolReader's suspend script (older one, 2015.7)

Only problem I have it's that device looked freezed for a moment one time (didn't changed screen message, nothing), I pressed button numerous times and when it unfreezed it started doing all power button presses I did (well, can't tell for sure, there were many, but I guess). Even after that cycle touch screen works just fine. This could be from using CR's suspend script, don't know.

**Comment by NiLuJe** [comment]
Okay, then limiting the workaround to weird people like me launching it from shell with nickel running was probably a good decision, thanks ;).

**Comment by Frenzie** [comment]
Oh, I do seem to recall an issue like that when I launched KOReader over
telnet while nickel was running. Not something I considered worth
investigating. My normal use is from KSM.

**Comment by NiLuJe** [comment]
I vaguely remember originally only having the issue in *nickel* after restarting it on KOReader's exit, FWIW.

But I think that one was originally fixed by properly killing the WiFi, so it may only be tangentially related.

But, yeah, if it only happens over shell w/ nickel running, good to confirm.

**Comment by Hzj-jie** [comment]
I have exactly the same firmware version for my kobo aura hd. I am using KoReader with KSM, and I have no this issue.
And I do not have /sys/devices/virtual/input/input1/neocmd on my device, it may be device specific.

**Comment by NiLuJe** [comment]
Huh, that's interesting, I though the Aura HD used a neonode IR grid, too?

**Comment by NiLuJe** [comment]
Hmm, code is there in the AuraHD kernel sources, though... Maybe the path is slightly different?

**Comment by houqp** [comment]
Same with KSM+KOReader in AuraHD. I never launch nickel aside with KOReader, so there is that :)

**Comment by Markismus** [comment]
@NiLuJe I upgraded to v.251 last night and I got the bug on my Kobo Aura H2O, too!

**Comment by Markismus** [comment]
I updated the firmware for Nickel, but the problem remains.

**Comment by NiLuJe** [comment]
Huh. Could the reordering of stuff and/or the added sleep be messing things up?

That's the only vaguely related commits I can think of in that window, I think?

**Comment by houqp** [comment]
@Markismus you upgraded from 175 to 251 right? Same question from NiluJe, do you have nickel running with KOReader? If you are only launching it from KSM, can you try manually reverting following two commits?
https://github.com/koreader/koreader/commit/ada373e247020c484053a09d76ffbd675b00a7e1
https://github.com/koreader/koreader/commit/767e6e2b7109ab97f9e9b8f6022abf15140f59cb

**Comment by Markismus** [comment]
@NiLuJe Yep, KSM.
@houqp Nope, I never run Nickel if I can help it. I hope I get to those revertions tomorrow, but it'll probably be later. I did revert back to v175 and the problem disappears. Which is odd, since NiLuJe suggested that he had the problem at v175, too.

**Comment by houqp** [comment]
Your case is probably a different from NiLuJe's since he is running it along side of nickel from SSH.

**Comment by NiLuJe** [comment]
Yep, and that (mine) specific scenario should trigger the workaround.
